google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java

```diff
@@ -146,24 +176,41 @@ public TransactionRunner readWriteTransaction() {
   public TransactionManager transactionManager() {
     Span span = tracer.spanBuilder(READ_WRITE_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
-      return pool.getReadWriteSession().transactionManager();
+      return getReadWriteSession().transactionManager();
     } catch (RuntimeException e) {
       TraceUtil.endSpanWithFailure(span, e);
       throw e;
     }
   }
 
   @Override
-  public long executePartitionedUpdate(Statement stmt) {
+  public long executePartitionedUpdate(final Statement stmt) {
     Span span = tracer.spanBuilder(PARTITION_DML_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
-      return pool.getReadWriteSession().executePartitionedUpdate(stmt);
+      return runWithSessionRetry(
+          new Function<Session, Long>() {
+            @Override
+            public Long apply(Session session) {
+              return session.executePartitionedUpdate(stmt);
+            }
+          });
     } catch (RuntimeException e) {
       TraceUtil.endSpanWithFailure(span, e);
       throw e;
     }
   }
 
+  private <T> T runWithSessionRetry(Function<Session, T> callable) {
+    PooledSession session = getReadWriteSession();
+    while (true) {
```

Comment:
Will this respect client side timeout values set ?
If our backend is down, and does not return any sessions -> whats the outcome of the customer application ?

Reply:
The short answer: It will behave the same as now, and the behavior will depend on whether the pool has a session ready to be returned or not. The `runWithSessionRetry` will only retry on the specific condition of a `SessionNotFoundException` being returned.

There are two possible scenarios:
1. The application requests a session and the session pool has a session that it thinks is valid. This session is handed to the application and the application tries to execute the transaction. This will cause the query or begin transaction statement to time out.
2. The pool does not have a session available that can be returned and starts the creation of a new session asynchronously. The requesting thread is placed in a waiter. The create session request will fail and notify the waiter of the error. The waiting thread will return with the create session error.

Both of the above will not be considered a `SessionNotFoundException` by the `runWithSessionRetry` method, and the method will return with an error.